torch_xla/csrc/tensor_methods.cpp

```diff
@@ -620,14 +620,31 @@ XLATensor XLATensor::cast(const XLATensor& input, at::ScalarType dtype) {
 
 XLATensor XLATensor::cat(tensorflow::gtl::ArraySlice<const XLATensor> tensors,
                          xla::int64 dim) {
+  // Shape checks for cat:
+  // - If not empty, every tensor shape must be the same.
+  // - Empty tensor passes but is simply ignore in implementation,
+  //   e.g. ([2, 3, 5], [])
+  // - If empty dimension, other dimensions must be the same.
+  //   e.g. ([4, 0, 32, 32], [4, 2, 32, 32], dim=1) passes.
+  //   ([4, 0, 32, 32], [4, 2, 31, 32], dim=1) throws.
   XLA_CHECK_GT(tensors.size(), 0);
   std::vector<ir::Value> values;
-  for (auto& tensor : tensors) {
-    if (xla::ShapeUtil::ElementsIn(tensor.shape()) > 0) {
-      dim = XlaHelpers::GetCanonicalDimensionIndex(dim,
-                                                   tensor.shape().get().rank());
-      values.push_back(tensor.GetIrValue());
+  std::vector<xla::Shape> shapes;
+  for (size_t i = 0; i < tensors.size(); ++i) {
+    xla::Shape tensor_shape = tensors[i].shape();
+    if (tensor_shape.rank() == 1 && tensor_shape.dimensions()[0] == 0) {
+      continue;
     }
+    dim = XlaHelpers::GetCanonicalDimensionIndex(dim, tensor_shape.rank());
+    tensor_shape.DeleteDimension(dim);
+    shapes.push_back(tensor_shape);
+    values.push_back(tensors[i].GetIrValue());
+  }
+  if (values.size() == 0) {
+    return tensors[0];
+  }
+  for (size_t i = 1; i < shapes.size(); ++i) {
```

Comment:
Why an extra tail loop here?
While you are pushing above, you can check.
```C++
if (!shapes.empty()) {
  XLA_CHECK(xla::ShapeUtil::Equal(shapes.back(), tensor_shape)) << shapes.back() << " vs " << tensor_shape;
}
shapes.push_back(tensor_shape);
```

Reply:
Nice catch! Updating...